libavcodec/nvjpegenc.c

```
@@ -0,0 +1,335 @@
+            subsampling = NVJPEG_CSS_410;
+            break;
+        default:
+            subsampling = NVJPEG_CSS_UNKNOWN;
+            break;
+        }
+        if (subsampling == NVJPEG_CSS_UNKNOWN) {
+            av_log(avctx, AV_LOG_ERROR, "Unsupported chroma subsampling: %d\n",
+                   ctx->subsampling);
+            return AVERROR(EINVAL);
+        }
+
+        CHECK_NVJPEG(nvjpegEncodeYUV(ctx->nvjpeg_handle, ctx->encoder_state,
+                                     ctx->encode_params, &nv_image, subsampling,
+                                     ctx->width, ctx->height, NULL));
+    } else {
+        // RGB/BGR format
+        CHECK_NVJPEG(nvjpegEncodeImage(ctx->nvjpeg_handle, ctx->encoder_state,
+                                       ctx->encode_params, &nv_image,
+                                       (nvjpegInputFormat_t)input_format,
+                                       ctx->width, ctx->height, NULL));
+    }
+
+    // Retrieve the bitstream and get the output buffer size
+    CHECK_NVJPEG(nvjpegEncodeRetrieveBitstream(
+        ctx->nvjpeg_handle, ctx->encoder_state, NULL, &out_buf_size, NULL));
+
+    out_buf = av_malloc(out_buf_size);
+    if (!out_buf) {
+        av_log(avctx, AV_LOG_ERROR, "Failed to allocate output buffer\n");
+        return AVERROR(ENOMEM);
+    }
+
+    // Retrieve the bitstream again to populate output buffer
+    CHECK_NVJPEG(nvjpegEncodeRetrieveBitstream(
+        ctx->nvjpeg_handle, ctx->encoder_state, out_buf, &out_buf_size, NULL));
+
+    ret = av_packet_from_data(pkt, out_buf, out_buf_size);
+    if (ret < 0) {
+        av_free(out_buf);
+        av_log(avctx, AV_LOG_ERROR, "av_packet_from_data failed\n");
+        return ret;
+    }
+
+    *got_packet = 1;
+    return 0;
+}
+
+static av_cold int nvjpeg_close(AVCodecContext* avctx) {
+    NvjpegContext* ctx = avctx->priv_data;
+
+    if (ctx->encoder_state) {
+        CHECK_NVJPEG(nvjpegEncoderStateDestroy(ctx->encoder_state));
```

Should we continue with cleanup even if one of these fails?

---

Should be safe for now. We'll emit a log if this does happen but once we get to this point we'd have a valid output so I don't think we should throw an error even if this fails. We should still try to destroy the rest of the states if possible anyway.

---

But CHECK_NVJPEG would return on error, right?

---

You're right. Removed it from all these and just manually emitting a log